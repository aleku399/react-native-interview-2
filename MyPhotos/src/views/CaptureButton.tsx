import React, {useRef} from 'react';
import {StyleSheet, View, ViewProps} from 'react-native';
import {TapGestureHandler} from 'react-native-gesture-handler';
import Reanimated, {
  Easing,
  useAnimatedStyle,
  withSpring,
  withTiming,
  useSharedValue,
  withRepeat,
} from 'react-native-reanimated';
import type {Camera} from 'react-native-vision-camera';
import {CAPTURE_BUTTON_SIZE} from './../Constants';

const BORDER_WIDTH = CAPTURE_BUTTON_SIZE * 0.1;

interface Props extends ViewProps {
  camera?: React.RefObject<Camera>;

  flash: 'off' | 'on';

  enabled: boolean;

  setIsPressingButton: (isPressingButton: boolean) => void;

  capturePhotos: () => void;
}

const _CaptureButton: React.FC<Props> = ({
  enabled,
  style,
  capturePhotos,
  ...props
}): React.ReactElement => {
  const isPressingButton = useSharedValue(false);

  const tapHandler = useRef<TapGestureHandler>();

  const shadowStyle = useAnimatedStyle(
    () => ({
      transform: [
        {
          scale: withSpring(isPressingButton.value ? 1 : 0, {
            mass: 1,
            damping: 35,
            stiffness: 300,
          }),
        },
      ],
    }),
    [isPressingButton],
  );

  const buttonStyle = useAnimatedStyle(() => {
    let scale: number;
    if (enabled) {
      if (isPressingButton.value) {
        scale = withRepeat(
          withSpring(1, {
            stiffness: 100,
            damping: 1000,
          }),
          -1,
          true,
        );
      } else {
        scale = withSpring(0.9, {
          stiffness: 500,
          damping: 300,
        });
      }
    } else {
      scale = withSpring(0.6, {
        stiffness: 500,
        damping: 300,
      });
    }

    return {
      opacity: withTiming(enabled ? 1 : 0.3, {
        duration: 100,
        easing: Easing.linear,
      }),
      transform: [
        {
          scale: scale,
        },
      ],
    };
  }, [enabled, isPressingButton]);

  return (
    <TapGestureHandler
      enabled={enabled}
      ref={tapHandler}
      onHandlerStateChange={capturePhotos}
      shouldCancelWhenOutside={false}
      maxDurationMs={99999999}>
      <Reanimated.View {...props} style={[buttonStyle, style]}>
        <Reanimated.View style={styles.flex}>
          <Reanimated.View style={[styles.shadow, shadowStyle]} />
          <View style={styles.button} />
        </Reanimated.View>
      </Reanimated.View>
    </TapGestureHandler>
  );
};

export const CaptureButton = React.memo(_CaptureButton);

const styles = StyleSheet.create({
  flex: {
    flex: 1,
  },
  shadow: {
    position: 'absolute',
    width: CAPTURE_BUTTON_SIZE,
    height: CAPTURE_BUTTON_SIZE,
    borderRadius: CAPTURE_BUTTON_SIZE / 2,
    backgroundColor: '#e34077',
  },
  button: {
    width: CAPTURE_BUTTON_SIZE,
    height: CAPTURE_BUTTON_SIZE,
    borderRadius: CAPTURE_BUTTON_SIZE / 2,
    borderWidth: BORDER_WIDTH,
    borderColor: 'white',
  },
});
