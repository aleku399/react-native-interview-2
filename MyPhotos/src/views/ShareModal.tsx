import {Text, Input, VStack} from 'native-base';
import React, {useState} from 'react';
import {StyleSheet, View, Alert, Modal} from 'react-native';
import {PressableOpacity} from 'react-native-pressable-opacity';
import IonIcon from 'react-native-vector-icons/Ionicons';

interface Props {
  showModal: boolean;
  closeModal: () => void;
  uploadImage: () => Promise<void>;
}

function ShareModal({showModal, closeModal, uploadImage}: Props) {
  const [endPoint, setEndPoint] = useState('');

  return (
    <Modal visible={showModal} transparent={true}>
      <View style={styles.container}>
        <View style={styles.main}>
          <PressableOpacity style={styles.closeButton} onPress={closeModal}>
            <IonIcon name="close" size={35} color="black" style={styles.icon} />
          </PressableOpacity>
          <VStack space={1}>
            <PressableOpacity onPress={uploadImage}>
              <Text style={styles.text}>Send to backend api</Text>
            </PressableOpacity>
            <VStack space={1}>
              <Text style={styles.text}>Enter api end point</Text>
              <Input
                mx={3}
                w="75%"
                maxWidth="300px"
                placeholder="Api End point"
                onChangeText={value => setEndPoint(value)}
                onBlur={() => Alert.alert(`Shared On ${endPoint}`)}
              />
            </VStack>
            <View style={styles.iconContainer}>
              <PressableOpacity
                style={styles.shareButton}
                onPress={() => Alert.alert('Shared on twitter')}>
                <IonIcon
                  name="logo-instagram"
                  size={35}
                  color="black"
                  style={styles.icon}
                />
              </PressableOpacity>
              <PressableOpacity
                style={styles.shareButton}
                onPress={() => Alert.alert('Shared on instagram')}>
                <IonIcon
                  name="logo-twitter"
                  size={35}
                  color="black"
                  style={styles.icon}
                />
              </PressableOpacity>
            </View>
          </VStack>
        </View>
      </View>
    </Modal>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  main: {
    backgroundColor: '#F0FFF0',
    justifyContent: 'center',
    borderRadius: 10,
    width: 250,
    height: 250,
  },
  text: {
    marginLeft: 12,
  },
  iconContainer: {
    alignItems: 'center',
  },
  closeButton: {
    position: 'absolute',
    top: 0,
    left: 0,
    width: 40,
    height: 40,
    zIndex: 100,
  },
  input: {
    borderColor: 'black',
    width: '100%',
    borderWidth: 1,
    borderRadius: 10,
    padding: 10,
    top: 5,
  },
  shareButton: {
    width: 40,
    height: 40,
  },
  icon: {
    textShadowColor: 'black',
    top: 8,
    textShadowOffset: {
      height: 0,
      width: 0,
    },
  },
});

export default ShareModal;
