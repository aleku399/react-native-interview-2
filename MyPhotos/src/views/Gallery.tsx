import React, {useMemo, useState} from 'react';
import {
  Dimensions,
  FlatList,
  Image,
  SafeAreaView,
  StatusBar,
  StyleSheet,
  View,
} from 'react-native';
import {PressableOpacity} from 'react-native-pressable-opacity';
import IonIcon from 'react-native-vector-icons/Ionicons';
import type {NativeStackScreenProps} from '@react-navigation/native-stack';
import {Routes} from '../Routes';
import ShareModal from './ShareModal';

type Props = NativeStackScreenProps<Routes, 'MediaPage'>;

const Item = ({item, onPress}: {item: any; onPress: any}) => (
  <View>
    <Image style={styles.image} source={{uri: item.uri}} />
    <View style={styles.main}>
      <PressableOpacity style={styles.shareButton} onPress={onPress}>
        <IonIcon
          name="share-social-outline"
          size={35}
          color="white"
          style={styles.icon}
        />
      </PressableOpacity>
    </View>
  </View>
);

const Gallery = ({route}: Props) => {
  const {data} = route.params;
  const [isShared, setIsShared] = useState(false);

  const sources = useMemo(() => {
    const uriCaptures = data.map(path => ({uri: `file://${path}`, id: path}));
    return uriCaptures;
  }, [data]);

  const renderItem = ({item}: {item: any}) => {
    return <Item item={item} onPress={() => setIsShared(true)} />;
  };

  return (
    <SafeAreaView style={styles.container}>
      <FlatList
        style={styles.flatListStyle}
        data={sources}
        renderItem={renderItem}
        numColumns={2}
        keyExtractor={item => item.id}
      />
      <ShareModal showModal={isShared} closeModal={() => setIsShared(false)} />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: StatusBar.currentHeight || 0,
    backgroundColor: 'black',
  },
  main: {
    flex: 1,
    alignItems: 'center',
  },
  item: {
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16,
  },
  title: {
    fontSize: 32,
  },
  image: {
    width: Dimensions.get('window').width / 2 - 20,
    height: 150,
    margin: 10,
  },
  flatListStyle: {flex: 1},
  shareButton: {
    width: 40,
    height: 40,
  },
  icon: {
    textShadowColor: 'black',
    textShadowOffset: {
      height: 0,
      width: 0,
    },
    textShadowRadius: 1,
  },
});

export default Gallery;
