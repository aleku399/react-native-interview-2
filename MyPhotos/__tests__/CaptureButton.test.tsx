import React from 'react';
import {act, create, ReactTestRenderer} from 'react-test-renderer';
import {CaptureButton} from '../src/views/CaptureButton';

it('renders the capture button correctly', () => {
  let root: ReactTestRenderer = create(<div />);

  const capturePhotos = jest.fn();

  act(() => {
    root = create(
      <CaptureButton
        flash={'off'}
        setIsPressingButton={jest.fn}
        enabled={false}
        capturePhotos={capturePhotos}
      />,
    );
  });

  expect(root.toJSON()).toMatchSnapshot();
});
