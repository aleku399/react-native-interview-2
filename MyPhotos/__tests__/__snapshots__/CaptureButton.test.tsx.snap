// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the capture button correctly 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {
        "opacity": 0.3,
        "transform": Array [
          Object {
            "scale": 0.6,
          },
        ],
      },
    }
  }
  collapsable={false}
  flash="off"
  forwardedRef={[Function]}
  onGestureHandlerEvent={[Function]}
  setIsPressingButton={[Function]}
  style={
    Object {
      "opacity": 0.3,
      "transform": Array [
        Object {
          "scale": 0.6,
        },
      ],
    }
  }
>
  <View
    animatedStyle={
      Object {
        "value": Object {},
      }
    }
    collapsable={false}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      animatedStyle={
        Object {
          "value": Object {
            "transform": Array [
              Object {
                "scale": 0,
              },
            ],
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#e34077",
          "borderRadius": 39,
          "height": 78,
          "position": "absolute",
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 78,
        }
      }
    />
    <View
      style={
        Object {
          "borderColor": "white",
          "borderRadius": 39,
          "borderWidth": 7.800000000000001,
          "height": 78,
          "width": 78,
        }
      }
    />
  </View>
</View>
`;
